src/log.js

```diff
@@ -85,46 +135,245 @@ export class Log {
+   * @return {T} The value of shouldBeTrueish.
+   * @template T
+   */
+  /*eslint "google-camelcase/google-camelcase": 0*/
+  assert(shouldBeTrueish, message, var_args) {
+    let firstElement;
+    if (!shouldBeTrueish) {
+      message = message || 'Assertion failed';
+      const splitMessage = message.split('%s');
+      const first = splitMessage.shift();
+      let formatted = first;
+      const messageArray = [];
+      pushIfNonEmpty(messageArray, first);
+      for (let i = 2; i < arguments.length; i++) {
+        const val = arguments[i];
+        if (val && val.tagName) {
+          firstElement = val;
+        }
+        const nextConstant = splitMessage.shift();
+        messageArray.push(val);
+        pushIfNonEmpty(messageArray, nextConstant.trim());
+        formatted += toString(val) + nextConstant;
+      }
+      const e = new Error(formatted);
+      e.fromAssert = true;
+      e.associatedElement = firstElement;
+      e.messageArray = messageArray;
+      this.prepareError_(e);
+      throw e;
+    }
+    return shouldBeTrueish;
+  }
+  /*eslint "google-camelcase/google-camelcase": 2*/
+
+  /**
+   * Asserts and returns the enum value. If the enum doesn't contain such a value,
+   * the error is thrown.
+   *
+   * @param {!Enum<T>} enumObj
+   * @param {string} s
+   * @param {string=} opt_enumName
+   * @return T
+   * @template T
+   */
+  assertEnumValue(enumObj, s, opt_enumName) {
+    for (const k in enumObj) {
+      if (enumObj[k] == s) {
+        return enumObj[k];
+      }
+    }
+    this.assert(false,
+        'Unknown %s value: "%s"',
+        opt_enumName || 'enum', s);
+  }
+
+  /**
+   * @param {!Error} error
+   * @private
+   */
+  prepareError_(error) {
+    if (this.suffix_) {
+      if (!error.message) {
+        error.message = this.suffix_;
+      } else if (error.message.indexOf(this.suffix_) == -1) {
+        error.message += this.suffix_;
+      }
+    }
+  }
+}
+
+
+/**
+ * @param {*} val
+ * @return {string}
+ */
+function toString(val) {
+  if (val instanceof Element) {
+    return val.tagName.toLowerCase() + (val.id ? '#' + val.id : '');
+  }
+  return val;
+}
+
+
+/**
+ * @param {!Array} array
+ * @param {*} val
+ */
+function pushIfNonEmpty(array, val) {
+  if (val != '') {
+    array.push(val);
+  }
+}
+
+
+/**
+ * @deprecated Use either publog or devlog
+ * TODO(dvoytenko, #2527): Remove this constant.
+ */
+export const log = new Log(window, mode => {
+  return mode.log == '1' ? LogLevel.FINE : LogLevel.OFF;
+});
+
+
+/**
+ * Publisher level log.
+ *
+ * Enabled in the following conditions:
+ *  1. Not disabled using `#log=0`.
+ *  2. Development mode is enabled via `#development=1` or logging is explicitly
+ *     enabled via `#log=D` where D >= 1.
+ *
+ * @const {!Log}
+ */
+export const user = new Log(window, mode => {
+  const logNum = parseInt(mode.log, 10);
+  if (mode.development || logNum >= 1) {
+    return LogLevel.FINE;
```

Comment:
Not super sure this is correct. But it is fine if we in practice only do `user.error` and `user.warn`